src/Spice86.Core/Emulator/Devices/Timer/Timer.cs

```diff
@@ -114,7 +111,7 @@ public class Timer : DefaultIOPortHandler {
         }
 
         if (_vgaScreenRefreshCounter.ProcessActivation(cycles)) {
-            _vgaCard?.UpdateScreen();
+            _machine.VideoSubsystem.VgaCard.UpdateScreen();
```

Comment:
This is the reverse of the direction I'd like to see.
Ideally these classes do **not** get a Machine injected, but only the things they depend on, through interfaces.
We want to put less in the Machine, not more.

Reply:
To depend on the VgaCard directly would produce a NullReferenceException later. At the time of initialization, the VgaCard does not exist yet.

Reply:
The way it was, a _reference_ to an `IVideoCard` was injected. As long as that reference pointed to nothing, nothing happened (due to the `?`). Once there _was_ something implementing that interface there, it had it's UpdateScreen method called.

The way you do it now the timer has to depend on a Machine which has to depend on a VideoSubSystem which has to depend on a VgaCard. All of which now have to exist and point to something, even though none of it is required by the Timer here. That only needs a reference to an `IVideoCard` to call UpdateScreen() on if it's filled.

The end goal is to have all the constructors parameters filled with only exactly the nbterfaces they themselves need.
And then have those be provided by StrongInject right?

Reply:
>  And then have those be provided by StrongInject right?

Yep. Although I'm not a fan of calling  the Resolve method of the container in our code.
This seems to be common with Source Generator based DI frameworks....

Putting everything (both emulator, UI, and other services) in one IContanier will be a nightmare to mock.

OTOH, using several containers we will use the Resolve

The code should be written as if This is what is done by user code in an ASP .NET Core webapp thanks to the reflection based magic of the Microsoft DI nugget package with which the user code first registers services (for each there is an interface and a set lifetime), and when a HTTP request hits a MVC controller, the Generic Host instantiates it itself. You don't have to Resolve services.

Maybe we could depend on the IContainer interface for the emulator container / UI container (depending on the layer) on the root object (for the UI or console, this is Program.cs).

But what is the root object for the emulator layer ? The ProgramExecutor ?
But the SoundBlaster for example will not be instantiated by the ProgramExecutor.
As of now, it is instantiated by the Machine. Should we resolve the container service there ?
But then the Machine ctor will depend on the IContainer interface, and then at this stage we might put the emulator IContainer as a dependcy for every class in the emulator assembly.
Or maybe for sound, have a SoundContainer... ?!

Yeah, I'm not a fan of the DI framework being apparent in the code. It should be transparent. 

AvaloniaUI's host both does not offer such a feature, and is incompatible with the Generic Host...
Plus we instantiate the emulator ourselves, we don't react to HTTP methods...

BTW, I corrected the Timer. It now depends directly on the IVideoCard.

Reply:
The trick with a DI framework is to never `new` anything that has an interface (which everything that has functionality should have).

`ProgramExecutor` needs a `Machine`? Inject an `IMachine` into its constructor.
`Machine` needs a `MachineCreationOptions`? Inject an `IMachineCreationOptions` into its constructor.
`MachineCreationOptions` needs an `IGui`, an `ILoggerService`, a `CounterConfigurator`, an `ExecutionFlowRecorder`, and a `Configuration`? Inject their interfaces into its constructor.
etc.

This makes the DI completely transparant. But it also requires a throrough understanding of all the (inter)dependencies in the code base, and elimination of cyclical dependencies, etc.

I feel this a bridge too far for this project at this stage.

Like I said before, we are currently using the `Machine` class as our dependency container.
Almost all dependencies are instantiated there, the order and requirements of the instantiations is taken care of there, the configuration of implementations, basically everything a dependency injection framework does is done there.
We also use it as a Service Locator. It is passed into almost all classes to provide a way for those classes to Get their dependencies from the Machine.

Rather than hiding this functionality deeper into "subsystems" I'd suggest splitting the Machine class in two : 1 "Machine" and 1 "ServiceContainer".
Isolating all the "Dependency Management/ Service Container" functionality makes it possible to maybe later on move it into a DI framework.
Or we may choose to keep our own homegrown ServiceContainer (this is now hidden inside the Machine) and build on it and improve it.

All this is not essential to the testability of the code. Interfaces are. Replacing every concrete (non-data-object) class with an interface in the constructor parameters will help immensly in making the code unit testable. And since the three most common parameters are `Machine`, `Configuration` and `ILoggerService` 

One approach would be to slowly replace the Machine(in its role of ServiceContainer) injection in the "leaves" of our codebase with the interfaces that the code actually uses. That way, usage of the Machine will move more and more to the "top" where it can eventually be replaced by something like StrongInject.

Reply:
I feel you are right.
But there is one thing I think we need to consider this is that devices lifecycle is naturally bound to the Machine.
You encouter HLT -> you dont only stop execution, you also reset all your devices.
Maybe execution part of the machine could be in another class?